File: backend/src/models/LichPhongVan.js
const mongoose = require('mongoose');
const { TRANG_THAI_LICH_PV, HINH_THUC_PHONG_VAN } = require('../utils/constants');

const lichPhongVanSchema = new mongoose.Schema({
  ungVienId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'UngVien',
    required: true
  },
  jobId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'TinTuyenDung',
    required: true
  },
  hoiDongId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'HoiDongPhongVan',
    required: true
  },
  thoiGianBatDau: {
    type: Date,
    required: [true, 'Vui lòng chọn thời gian bắt đầu']
  },
  thoiGianKetThuc: {
    type: Date,
    required: [true, 'Vui lòng chọn thời gian kết thúc']
  },
  diaDiem: {
    type: String,
    trim: true
  },
  linkOnline: {
    type: String,
    trim: true
  },
  hinhThuc: {
    type: String,
    enum: Object.values(HINH_THUC_PHONG_VAN),
    required: true
  },
  trangThai: {
    type: String,
    enum: Object.values(TRANG_THAI_LICH_PV),
    default: TRANG_THAI_LICH_PV.CHO_XAC_NHAN
  },
  noiDung: {
    type: String,
    trim: true
  },
  ghiChu: {
    type: String,
    trim: true
  },
  taoBoi: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung',
    required: true
  },
  ngayXacNhan: {
    type: Date
  },
  nguoiXacNhan: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung'
  },
  lyDoTuChoi: {
    type: String,
    trim: true
  },
  lyDoHuy: {
    type: String,
    trim: true
  },
  nguoiHuy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NguoiDung'
  },
  ngayHuy: {
    type: Date
  },
  emailDaGui: {
    type: Boolean,
    default: false
  },
  tokenXacNhan: {
    type: String,
    unique: true,
    sparse: true
  }
}, {
  timestamps: true,
  collection: 'lich_phong_van'
});

// Index
lichPhongVanSchema.index({ ungVienId: 1 });
lichPhongVanSchema.index({ jobId: 1 });
lichPhongVanSchema.index({ hoiDongId: 1 });
lichPhongVanSchema.index({ thoiGianBatDau: 1 });
lichPhongVanSchema.index({ trangThai: 1 });
lichPhongVanSchema.index({ tokenXacNhan: 1 });

// Validate thời gian
lichPhongVanSchema.pre('save', function(next) {
  if (this.thoiGianKetThuc <= this.thoiGianBatDau) {
    return next(new Error('Thời gian kết thúc phải sau thời gian bắt đầu'));
  }
  
  // Validate địa điểm hoặc link online
  if (this.hinhThuc === HINH_THUC_PHONG_VAN.TRUC_TIEP && !this.diaDiem) {
    return next(new Error('Vui lòng nhập địa điểm phỏng vấn'));
  }
  
  if (this.hinhThuc === HINH_THUC_PHONG_VAN.ONLINE && !this.linkOnline) {
    return next(new Error('Vui lòng nhập link phỏng vấn online'));
  }
  
  next();
});

// Virtual để kiểm tra đã qua
lichPhongVanSchema.virtual('daQua').get(function() {
  return new Date() > this.thoiGianKetThuc;
});

// Virtual để tính thời lượng (phút)
lichPhongVanSchema.virtual('thoiLuong').get(function() {
  const diff = this.thoiGianKetThuc - this.thoiGianBatDau;
  return Math.round(diff / (1000 * 60));
});

// Method xác nhận lịch
lichPhongVanSchema.methods.xacNhan = function(nguoiXacNhanId) {
  this.trangThai = TRANG_THAI_LICH_PV.DA_XAC_NHAN;
  this.ngayXacNhan = new Date();
  this.nguoiXacNhan = nguoiXacNhanId;
};

// Method từ chối lịch
lichPhongVanSchema.methods.tuChoi = function(lyDo) {
  this.trangThai = TRANG_THAI_LICH_PV.TU_CHOI;
  this.lyDoTuChoi = lyDo;
};

// Method hủy lịch
lichPhongVanSchema.methods.huy = function(nguoiHuyId, lyDo) {
  this.trangThai = TRANG_THAI_LICH_PV.HUY;
  this.nguoiHuy = nguoiHuyId;
  this.ngayHuy = new Date();
  this.lyDoHuy = lyDo;
};

const LichPhongVan = mongoose.model('LichPhongVan', lichPhongVanSchema);

module.exports = LichPhongVan;

File: frontend/lib/models/lich_phong_van.dart
class LichPhongVan {
  final String id;
  final String ungVienId;
  final String jobId;
  final String hoiDongId;
  final DateTime thoiGianBatDau;
  final DateTime thoiGianKetThuc;
  final String? diaDiem;
  final String? linkOnline;
  final String hinhThuc;
  final String trangThai;
  final String? noiDung;
  final String? ghiChu;
  final String taoBoi;
  final DateTime? ngayXacNhan;
  final String? nguoiXacNhan;
  final String? lyDoTuChoi;
  final String? lyDoHuy;
  final String? nguoiHuy;
  final DateTime? ngayHuy;
  final bool emailDaGui;
  final String? tokenXacNhan;
  final DateTime createdAt;
  final DateTime updatedAt;

  LichPhongVan({
    required this.id,
    required this.ungVienId,
    required this.jobId,
    required this.hoiDongId,
    required this.thoiGianBatDau,
    required this.thoiGianKetThuc,
    this.diaDiem,
    this.linkOnline,
    required this.hinhThuc,
    required this.trangThai,
    this.noiDung,
    this.ghiChu,
    required this.taoBoi,
    this.ngayXacNhan,
    this.nguoiXacNhan,
    this.lyDoTuChoi,
    this.lyDoHuy,
    this.nguoiHuy,
    this.ngayHuy,
    required this.emailDaGui,
    this.tokenXacNhan,
    required this.createdAt,
    required this.updatedAt,
  });

  factory LichPhongVan.fromJson(Map<String, dynamic> json) {
    return LichPhongVan(
      id: json['_id'] ?? '',
      ungVienId: json['ungVienId'] is String 
          ? json['ungVienId'] 
          : json['ungVienId']?['_id'] ?? '',
      jobId: json['jobId'] is String 
          ? json['jobId'] 
          : json['jobId']?['_id'] ?? '',
      hoiDongId: json['hoiDongId'] is String 
          ? json['hoiDongId'] 
          : json['hoiDongId']?['_id'] ?? '',
      thoiGianBatDau: DateTime.parse(json['thoiGianBatDau']),
      thoiGianKetThuc: DateTime.parse(json['thoiGianKetThuc']),
      diaDiem: json['diaDiem'],
      linkOnline: json['linkOnline'],
      hinhThuc: json['hinhThuc'] ?? 'Truc_tiep',
      trangThai: json['trangThai'] ?? 'Cho_xac_nhan',
      noiDung: json['noiDung'],
      ghiChu: json['ghiChu'],
      taoBoi: json['taoBoi'] ?? '',
      ngayXacNhan: json['ngayXacNhan'] != null 
          ? DateTime.parse(json['ngayXacNhan']) 
          : null,
      nguoiXacNhan: json['nguoiXacNhan'],
      lyDoTuChoi: json['lyDoTuChoi'],
      lyDoHuy: json['lyDoHuy'],
      nguoiHuy: json['nguoiHuy'],
      ngayHuy: json['ngayHuy'] != null 
          ? DateTime.parse(json['ngayHuy']) 
          : null,
      emailDaGui: json['emailDaGui'] ?? false,
      tokenXacNhan: json['tokenXacNhan'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'ungVienId': ungVienId,
      'jobId': jobId,
      'hoiDongId': hoiDongId,
      'thoiGianBatDau': thoiGianBatDau.toIso8601String(),
      'thoiGianKetThuc': thoiGianKetThuc.toIso8601String(),
      'diaDiem': diaDiem,
      'linkOnline': linkOnline,
      'hinhThuc': hinhThuc,
      'trangThai': trangThai,
      'noiDung': noiDung,
      'ghiChu': ghiChu,
      'taoBoi': taoBoi,
      'ngayXacNhan': ngayXacNhan?.toIso8601String(),
      'nguoiXacNhan': nguoiXacNhan,
      'lyDoTuChoi': lyDoTuChoi,
      'lyDoHuy': lyDoHuy,
      'nguoiHuy': nguoiHuy,
      'ngayHuy': ngayHuy?.toIso8601String(),
      'emailDaGui': emailDaGui,
      'tokenXacNhan': tokenXacNhan,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  // Helper methods
  bool get daQua => DateTime.now().isAfter(thoiGianKetThuc);
  bool get sapDienRa => DateTime.now().isBefore(thoiGianBatDau) && 
                        thoiGianBatDau.difference(DateTime.now()).inHours <= 24;
  
  int get thoiLuongPhut => thoiGianKetThuc.difference(thoiGianBatDau).inMinutes;
  
  String get hinhThucText {
    const map = {
      'Truc_tiep': 'Trực tiếp',
      'Online': 'Online',
      'Dien_thoai': 'Điện thoại',
    };
    return map[hinhThuc] ?? hinhThuc;
  }
  
  String get trangThaiText {
    const map = {
      'Cho_xac_nhan': 'Chờ xác nhận',
      'Da_xac_nhan': 'Đã xác nhận',
      'Tu_choi': 'Từ chối',
      'Hoan_thanh': 'Hoàn thành',
      'Huy': 'Hủy',
    };
    return map[trangThai] ?? trangThai;
  }
}